core/server/worker/src/main/java/alluxio/worker/block/CacheRequestManager.java

```diff
@@ -257,11 +298,92 @@ public void run() {
         } else {
           ASYNC_CACHE_FAILED_BLOCKS.inc();
         }
-        mPendingRequests.remove(blockId);
+        mPendingAsyncRequests.remove(blockId);
       }
     }
   }
+  /**
+   * CacheTask is a callable task that can be considered equal if
+   * the blockId of the request is the same.
+   */
+  @VisibleForTesting
+  class CacheTask implements Callable<Boolean> {
+    private final CacheRequest mRequest;
+
+    /**
+     * Constructor for an AsyncCacheTask.
+     *
+     * @param request an AsyncCacheRequest
+     */
+    CacheTask(CacheRequest request) {
+      mRequest = request;
+    }
+
+    @Override
+    public int hashCode() {
+      // Only care about the block id being the same.
+      // Do not care if the source host or port is different.
+      return Objects.hash(mRequest.getBlockId());
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (this == obj) {
+        return true;
+      }
+      if (!(obj instanceof CacheRequestManager.CacheTask)) {
+        return false;
+      }
+      CacheRequestManager.CacheTask that = ((CacheRequestManager.CacheTask) obj);
+      if (mRequest == that.mRequest) {
+        return true;
+      }
+      if (this.mRequest == null || that.mRequest == null) {
+        return false;
+      }
+      // Only care about the block id being the same.
+      // Do not care if the source host or port is different.
+      return mRequest.getBlockId() == that.mRequest.getBlockId();
+    }
 
+    public Boolean call() {
+      boolean result = false;
+      long blockId = mRequest.getBlockId();
+      long blockLength = mRequest.getLength();
+      try {
+        boolean isSourceLocal = mLocalWorkerHostname.equals(mRequest.getSourceHost());
+        // Check if the block has already been cached on this worker
+        if (mBlockWorker.hasBlockMeta(mRequest.getBlockId())) {
+          CACHE_DUPLICATE_REQUESTS.inc();
```

Comment:
I don't think we need these metric counter to bump, it is not very useful.

Reply:
just through debug message for duplicate requests

Reply:
same for async variation.